### DISCUSSION TITLE
Help with API history

### DISCUSSION THREAD

**Discussion by TheChickenMan69** [oobabooga/text-generation-webui#2603] [Q&A] [answered]
does anyone understand how to use the new API history? I assume we still have to handle it ourselves, but how is the data supposed to be saved? I see in the api-example-chat script you have this:

```
# Basic example
    history = {'internal': [], 'visible': []}
```

And I see you pass it in with the request:

```
def run(user_input, history):
    request = {
        'user_input': user_input,
        'history': history,

```

but what info is suppose to be in internal and what is visible? do I just store bots messages in one and users in another or what? haha!

Any help would be greatly appreciated... and thank you in advance! 😄 Bonus points if you can show me some example code of storing the data!

**Comment by w0lfzk1n** [comment, marked as answer]
Hello there!
I had the same struggle the last couple of days and I was not able to find anything in the internet for this. So I did some debugging and testing, while throwing my laptop at anything and anyone.
Until i figured it out.
Here is an example, in javascript, how to use the History for the API correctly and working :D 

I hope it helps <3

=============================================================================
```
const axios = require('axios');
const readline = require('readline');

const HOST = 'localhost:5000';
const URI = `http://${HOST}/api/v1/chat`;

// Important thing for the AI to remember and use the history correctly!
// The Language Model will always work on the second string, in the last object of the history.internal or *.visible object.
// Therefore, when a new "human" prompt is entered, before making the API request, you need to push the user_input
// and a empty string (or the string containing the start of the AI's reply) which the AI will then use for the generation of it's prompt.

// It took me some time until i have figured it out, since there isn't much information about it's actual usage on the web.
// Or i wasn't able to find anything, so I thought i share my knowledge, maybe helping out some of you, struggling with this.

/*EXAMPLE HISTORY WITH A NEW USERINPUT BEFORE SENDING WITH API 

'''
let history = {
    internal: [ ["Hi there", "Hello, I am an AI. How can i help you?"], ["Please explain how language models work", ""] ],
    visible: [ ["Hi there",v"Hello, I am an AI. How can i help you?"], ["Please explain how language models work", ""] ]
}
'''

Optional you can here for example use "Language models are" for giving a emtpy "" string for the AI to work on.

*/

let history = {
    internal: [],
    visible: []
}

// API Request function
async function run(user_input, history) {
    const request = {
        user_input: user_input,
        history: history,
        mode: 'chat',  // Options: 'chat', 'chat-instruct', 'instruct'
        character: "Example",
        instruction_template: 'vicuna-v1.1',
        your_name: 'Human',
        regenerate: false,
        _continue: true, // Option for sending History
        stop_at_newline: false,
        chat_prompt_size: 2048,
        chat_generation_attempts: 1,
        chat_instruct_command: 'Continue the chat dialogue below. Write a single reply for the character "Assistant"\n\n',
        max_new_tokens: 250,
        do_sample: true,
        temperature: 0.7,
        top_p: 0.1,
        typical_p: 1,
        epsilon_cutoff: 0,
        eta_cutoff: 0,
        tfs: 1,
        top_a: 0,
        repetition_penalty: 1.18,
        top_k: 40,
        min_length: 0,
        no_repeat_ngram_size: 0,
        num_beams: 1,
        penalty_alpha: 0,
        length_penalty: 1,
        early_stopping: false,
        mirostat_mode: 0,
        mirostat_tau: 5,
        mirostat_eta: 0.1,
        seed: -1,
        add_bos_token: true,
        truncation_length: 2048,
        ban_eos_token: false,
        skip_special_tokens: true,
        stopping_strings: ["\n### Assistant:", "\n### Human:"] // I added these, because it seems like, all AI love Assistant and Human
    };

    try {
        console.log(">> SENDING API REQUEST")
        const response = await axios.post(URI, request);

        if (response.status === 200) {
            // Uncommend below, to show full API response
            //console.log("RESPONSE: ", response)

            // The "empty" string that now contains the AI's response
            const result = response.data.results[response.data.results.length - 1].history;
            console.log("Received AI answer: " + result.visible[result.visible.length - 1][1])

            // Set the empty string, that we first pushed for the AI to use for generating it's prompt, to the now received answer, in ["internal"] and ["visible"]
            history.internal[history.internal.length - 1][1] = result.visible[result.visible.length - 1][1]
            history.visible[history.internal.length - 1][1] = result.visible[result.visible.length - 1][1]

            // Uncomment below, to show full history
            //console.log("> NEW HISTORY:\n", history)

            return result.visible[result.visible.length - 1][1]

        } else {
            console.log(` ERROR IN RESPONSE:\n\n\n${response}`)
        }
    } catch (error) {
        console.error('ERROR WHEN SENDING API REQUEST:', error);
    }
}

// Javascriptpart to read the userinput in this example
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

// Loop 10x for a small conversation with the AI
for (i = 0; i < 10; i++) {
    // Call rl to read userinput
    rl.question(`[${i}] Your Input: `, (input) => {
        console.log('Human:', input);

        // Push a new conversation to the history, but with an empty string (or containing the start of AI reply) for the AI to generate on.
        history.internal.push([input, ""])
        history.visible.push([input, ""])

        console.log

        // Send API request with history.
        run(input, history)
        rl.close();
    });
}
```

**Reply by Sousheyyy** [reply]
Somehow API returns empty responses. It reaches to the server but can't get any answers back? Do you have any idea how to solve it? I'm using the same code here and an uploaded character.

**Reply by w0lfzk1n** [reply]
@Sousheyyy
Did you put the exact name of the character in the API request?
If the Character is named
"Xy3"
It should be
"Xy3"
In the request too.

Maybe try another character and see if that works, example the
"Example"
Character.

Maybe share some more infos about your issue.

**Comment by Jewbby** [comment]
So I used this which was a huge help in getting it to finally store and continue context. Though, any reason you can think of why it might start using emojis all of a sudden? Specifically it is responding only in Unicode. Either they're emojis which do seem to be related in what I'm saying or it'll be blank with \u200b. This is so bizarre.

**Reply by Jewbby** [reply]
So, I found the culprit of the unicode issue that was giving emojis or blank space. It seems by having set _continue: True, this somehow completely breaks it. Flipping it back to false and implementing a lot of your other suggestions does seem to help with storing and building context size unlike before. However, something simple like telling it who my favorite president is, then immediately asking it who my favorite president is, gives a random answer. So I guess I might need more tweaking on formatting the history. I'm not sure if _continue is specifically an issue with Python or not, but will not seem to work with it enabled.

**Reply by cureator** [reply]
Im in the same boat and was hoping for clarification too.  Is there anywhere more active people are discussing this? Hard to find info or a community around this api

**Reply by w0lfzk1n** [reply]
@Jewbby
Oh wow
I have never experienced my model using emojis.
I dont think its an issue with the _continue option.

It might be a issue with your model or model settings like temperatur or the model_template?

**Reply by sugarcanefarmer** [reply]
It's not with _continue for memory, just spent the night debugging. I legit thought the api was broken - it (mostly) isn't. Running on python. Here's what I picked up on...

1. 'chat-instruct' is broken. Maybe because I failed to add in an instruction in my request? (It says it will auto-pickup it though so...) But it was having issues either way. I swapped to 'chat' and then - boom - memory issues went away entirely (apart from, you know, normal AI issues haha). The character also had issues where they would basically ignore my text too; they literally would re-introduce themselves to me mid conversation. Again, setting it to just "chat" fixed that for me. EDIT: Oh yeah I also had issues with blank replies only some of the time... I think this fixed it, but I also just wrote something to keep requesting if a reply is blank so...

2. Yep, you gotta store the histories as they come in. Otherwise they won't remember anything. And send it in the request.

3. _continue does create emojis or blank space at the start of messages, sometimes both. I have no clue why this happens... I am just as confounded. The emojis were all related though, and I considered just leaving it so my character could emote. I end up leave it out of my request though; sending a message with just the previous history & new text as "user_input" works just fine, "_continue" is not needed, as well as creating the history with your message and a blank reply; neither are needed. That let me write out the code a bit more simply, just storing history after getting a reply using:
`history['internal'].append([user_input, received_message])
history['visible'].append([user_input, received_message])
`

4. I'm not sure if this helped, but I noticed python was storing text with single quotes sometimes. Not sure if the interpreter for the ooba API has issues with it... you know, saving history as 'Dad said to get milk' or "Dad said to get milk" might have been causing some issues? Default ooba convos export it all as "Double Quotes". It would just pick either single or double quotes, for both the AI and Me, at pretty much random. I ended up "cleaning" it by writing to and then reading from a .json for the history, and now it is all in "double quotes". The request is also sent as JSON, so like probably un-needed... but who knows how python works half the time. Plus it can save logs continually as a side effect, so win-win.

Sorry if this is a necro-post, but I don't want anyone else to do what I was about to do. I was boutta make a whole package using Playwright that would just use the web interface and not even the API lol. Good thing it's actually working.
